In [ ]:
# CT image is input

# INPUT: CT Images is input
# OUTPUT: RP file
# PROCESS: algorith predict by each slice. make_lines_process. wrap to RP file by RP file template

# Template processing note: 


# Some Filepath for RP 


In [ ]:
#rp_info =dicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')
#rp_info = pydicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')

# Template Data
rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')


# Only Read File

In [36]:

# Start research from here

import pydicom

#rp_info =dicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')
#rp_info = pydicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')

# Template Data
rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')

App = rp_info.ApplicationSetupSequence
ChannelSequence = App[0].ChannelSequence
#print('ChannelSequence[0].BrachyControlPointSequence = ')
#print(ChannelSequence[0].BrachyControlPointSequence)
print('ChannelSequence[0] = ')

for idx in range(len(ChannelSequence)):
    cs = ChannelSequence[idx]
    print('idx = {} ->'.format(idx))
    print(cs)
    break
    

for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):
    
    #print(cp)
    #print("position:", cp.ControlPoint3DPosition, end = " ")
    #print("position:", cp.ControlPoint3DPosition)
    #print(type(cp.ControlPoint3DPosition))
    
    if (idx >= 0 ) :
        print('idx == {} -> cp = '.format(idx))
        print(cp)
        print(type(cp.ControlPointIndex))
        print('===============')
    
    #continue
    print("Time:", cp.CumulativeTimeWeight)
    try: 
        for jj in cp.BrachyReferencedDoseReferenceSequence:
            #print(jj)
            print("ref:", rp_info.DoseReferenceSequence[jj.ReferencedDoseReferenceNumber-1].DoseReferenceDescription , end = " ")
            print("coeff:", jj.CumulativeDoseReferenceCoefficient)
    except Exception as e:
        print(e)
        pass
    print('')

ChannelSequence[0] = 
idx = 0 ->
(3006, 0084) Referenced ROI Number               IS: "18"
(300a, 0110) Number of Control Points            IS: "38"
(300a, 0282) Channel Number                      IS: "1"
(300a, 0284) Channel Length                      DS: "1300"
(300a, 0286) Channel Total Time                  DS: "0"
(300a, 0288) Source Movement Type                CS: 'STEPWISE'
(300a, 0290) Source Applicator Number            IS: "1"
(300a, 0291) Source Applicator ID                SH: 'Tandom_2'
(300a, 0292) Source Applicator Type              CS: 'RIGID'
(300a, 0294) Source Applicator Name              LO: ''
(300a, 0296) Source Applicator Length            DS: "1300"
(300a, 02a0) Source Applicator Step Size         DS: "5"
(300a, 02a2) Transfer Tube Number                IS: ''
(300a, 02c8) Final Cumulative Time Weight        DS: "0"
(300a, 02d0)  Brachy Control Point Sequence   38 item(s) ---- 
   (300a, 0112) Control Point Index                 IS: "0"
   (300a, 02d2) Contro

In [40]:
ChannelSequence[3]

IndexError: list index out of range

# Write File and read file again 

In [25]:

# Write File Test and read again 

import pydicom

#rp_info =dicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')
#rp_info = pydicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')

# Template Data
rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')
out_rp_filepath = r'Brachy.RP.out.dcm'

App = rp_info.ApplicationSetupSequence
ChannelSequence = App[0].ChannelSequence
print('ChannelSequence[0] = ')

for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):
    print("position:", cp.ControlPoint3DPosition)
    if (idx >= 0 ) :
        #print('idx == {} -> cp = '.format(idx))
        #print(cp)
        #print('===============')
        pass
    if (idx == 0) :
        cp.ControlPoint3DPosition[0] = 33.33333
    
        
    
    #continue
    print("Time:", cp.CumulativeTimeWeight)
    try: 
        for jj in cp.BrachyReferencedDoseReferenceSequence:
            #print(jj)
            print("ref:", rp_info.DoseReferenceSequence[jj.ReferencedDoseReferenceNumber-1].DoseReferenceDescription , end = " ")
            print("coeff:", jj.CumulativeDoseReferenceCoefficient)
    except Exception as e:
        print(e)
        pass
    print('')
pydicom.write_file(out_rp_filepath, rp_info)


in_rp_filepath = r'Brachy.RP.out.dcm'
rp_fp = pydicom.read_file(in_rp_filepath)
seq = rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence
for idx, cp in enumerate(seq):
    print("position:", cp.ControlPoint3DPosition)


ChannelSequence[0] = 
position: ['12.778672628366', '-14.486222523329', '1.80953906063796']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.778672628366', '-14.486222523329', '1.80953906063796']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.4938084644467', '-12.702280882149', '-8.025927609987']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.4938084644467', '-12.702280882149', '-8.025927609987']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.351376382487', '-11.810310061558', '-12.943

# The format is like 
rp_info.ApplicationSetupSequence[0].ChannelSequence[0].NumberOfControlPoints is 38
rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence[ 0 .. 37]



(300a, 0112) Control Point Index                 IS: "0"
(300a, 02d2) Control Point Relative Position     DS: "3.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.778672628366', '-14.486222523329', '1.80953906063796']
(300a, 02d6) Cumulative Time Weight              DS: "0"

(300a, 0112) Control Point Index                 IS: "1"
(300a, 02d2) Control Point Relative Position     DS: "3.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.778672628366', '-14.486222523329', '1.80953906063796']
(300a, 02d6) Cumulative Time Weight              DS: "0"


(300a, 0112) Control Point Index                 IS: "2"
(300a, 02d2) Control Point Relative Position     DS: "8.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
(300a, 02d6) Cumulative Time Weight              DS: "0"


(300a, 0112) Control Point Index                 IS: "3"
(300a, 02d2) Control Point Relative Position     DS: "8.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
(300a, 02d6) Cumulative Time Weight              DS: "0"

...

(300a, 0112) Control Point Index                 IS: "36"
(300a, 02d2) Control Point Relative Position     DS: "93.5"
(300a, 02d4) Control Point 3D Position           DS: ['10.2148951530921', '1.56925224729876', '-86.709660974987']
(300a, 02d6) Cumulative Time Weight              DS: "0"


(300a, 0112) Control Point Index                 IS: "37"
(300a, 02d2) Control Point Relative Position     DS: "93.5"
(300a, 02d4) Control Point 3D Position           DS: ['10.2148951530921', '1.56925224729876', '-86.709660974987']
(300a, 02d6) Cumulative Time Weight              DS: "0"

In [76]:
import pydicom
import copy

rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')


#App = rp_info.ApplicationSetupSequence
#ChannelSequence = App[0].ChannelSequence.Number of Control Points

print(rp_info.ApplicationSetupSequence[0].ChannelSequence[0].NumberOfControlPoints)
#print(rp_info.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence)

# We call BrachyControlPointSequence[i] BCPItem here
BCPItemTemplate = copy.deepcopy(rp_info.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence[37])
#print(BrachyControlPointSequenceItemTemplate)

#
# (ControlPointIndex, ControlPointRelativePosition, ControlPoint3DPosition, CumulativeTimeWeight)
# (0, 3.5, [,,], 0 )
# (0, )

#for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):

print(rp_info)

38
(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20191015'
(0008, 0013) Instance Creation Time              TM: '164400.648000'
(0008, 0016) SOP Class UID                       UI: RT Plan Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.5.417454940236.2063186.20191015164204
(0008, 0020) Study Date                          DA: '20191015'
(0008, 0030) Study Time                          TM: '103307.760000'
(0008, 0050) Accession Number                    SH: '2019-10-15-10:36'
(0008, 0060) Modality                            CS: 'RTPLAN'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Plans'
(0008, 1

## Write related RS Data in RP Tamplate

### OperatorsName 
cylin is in template. But Who is the people ?

### PhysiciansOfRecord
Just copy from RS

### FrameOfReferenceUID
In RS, ReferencedFrameOfReferenceSequence[0].FrameOfReferenceUID


### Basic thing copy from RS
PhysiciansOfRecord, PatientName, PatientID, PatientBirthDate, PatientBirthTime, PatientSex,
DeviceSerialNumber, SoftwareVersions, StudyID, StudyDate, StudyTime, StudyInstanceUID

### ReferencedStructureSetSequence
In RP,   
ReferencedStructureSetSequence[0].ReferencedSOPClassUID = 1.2.840.10008.5.1.4.1.1.481.3  ReferencedStructureSetSequence[0].ReferencedSOPInstanceUID =   1.2.246.352.71.4.417454940236.287582.20191015164114  
So,   
RP ReferencedStructureSetSequence[0].ReferencedSOPClassUID =  RS SOPClassUID  
RP ReferencedStructureSetSequence[0].ReferencedSOPInstanceUID =  RS SOPInstanceUID  

### InstanceCreation
RS InstanceCreationDate = 20191015  
RS InstanceCreationTime = 164359.971000  
RP InstanceCreationDate = 20191015  
RP InstanceCreationTime = 164400.648000  
RP RTPlanDate = 20191015  
RP RTPlanTime = 164203.870000   (What if compare to RP InstanceCreationTime ?)    
RS StudyDate = 20191015   
RS StudyTime = 103307.760000  
  
Conclusion,     
StudyDate, InstanceCreationDate are all the same in RS and RP.    
RS StudyTime < RP RTPlanTime < RS InstanceCreationTime < RP InstanceCreationTime   



# Implemnet write procedure

In [53]:
import pydicom
import os


def get_new_uid(old_uid='1.2.246.352.71.5.417454940236.2063186.20191015164204', study_date='20190923'):
    def gen_6_random_digits():
        import random
        ret_str = ""
        for i in range(6):
            ch = chr(random.randrange(ord('0'), ord('9') + 1))
            ret_str += ch
        return ret_str

    theStudyDate = study_date
    uid_last_part = uid.split('.')[-1] #20191015164204  first 8 digit is match to StudyDate
    uid_list = uid.split('.')
    uid_list[-1] = theStudyDate + gen_6_random_digits()
    #print(uid_list)
    new_uid = '.'.join(uid_list)
    #print(uid)
    #print(new_uid)
    return new_uid



rp_template_filepath = r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm'
#folder = r'RAL_plan_new_20190905/29059811-1'
#folder = r'RP_Template_TestData'
folder = r'RALmilo'

# Set rs_filepath and ct_filelist, which are INPUT
rs_filepath = ''
ct_filelist = []
for file in os.listdir(folder):
    filepath = os.path.join(folder, file)
    fp = pydicom.read_file(filepath)
    if (fp.Modality == 'CT'):
        ct_filelist.append(filepath)
    elif (fp.Modality == 'RTSTRUCT'):
        rs_filepath = filepath
        
# Read RS file as input
rs_fp = pydicom.read_file(rs_filepath)
# read RP tempalte into rp_fp
rp_fp = pydicom.read_file(rp_template_filepath)
out_rp_filepath = r'out.brachy.rp.dcm'


rp_fp.OperatorsName = 'cylin'
rp_fp.PhysiciansOfRecord = rs_fp.PhysiciansOfRecord
rp_fp.FrameOfReferenceUID = rs_fp.ReferencedFrameOfReferenceSequence[0].FrameOfReferenceUID
rp_fp.ReferencedStructureSetSequence[0].ReferencedSOPClassUID =  rs_fp.SOPClassUID  
rp_fp.ReferencedStructureSetSequence[0].ReferencedSOPInstanceUID = rs_fp.SOPInstanceUID  

directAttrSet = [
    'PhysiciansOfRecord','PatientName', 'PatientID', 
    'PatientBirthDate', 'PatientBirthTime', 'PatientSex', 
    'DeviceSerialNumber', 'SoftwareVersions', 'StudyID', 
    'StudyDate', 'StudyTime', 'StudyInstanceUID']
for attr in directAttrSet:
    rs_val = getattr(rs_fp, attr)
    rp_val = getattr(rp_fp, attr)
    print('attr={}, \n In RS->{} \n In RP->{}'.format(attr, rs_val, rp_val))
    
    val = getattr(rs_fp, attr)
    setattr(rp_fp, attr, val)
    new_rp_val = getattr(rp_fp,attr)
    print('after update, RP->{}\n'.format(new_rp_val))

newSeriesInstanceUID = get_new_uid(old_uid=rp_fp.SeriesInstanceUID, study_date=rp_fp.StudyDate )
newSOPInstanceUID = get_new_uid(old_uid=rp_fp.SOPInstanceUID, study_date=rp_fp.StudyDate)
rp_fp.SeriesInstanceUID = newSeriesInstanceUID
rp_fp.SOPInstanceUID = newSOPInstanceUID

    
# Change SeriesInstanceUID and SOPInstanceUID


# 
# to satisfy the condition 
# (1) RS InstanceCreationDate == RP InstanceCreationDate == RP RTPlanDate == RS StudyDate
# (2) RS StudyTime < RP RTPlanTime < RS InstanceCreationTime < RP InstanceCreationTime   
# for (1)
rp_fp.InstanceCreationDate = rp_fp.RTPlanDate = rp_fp.StudyDate = rs_fp.StudyDate
# for (2)
#print('type(rs_fp.StudyTime) = {}'.format(type(rs_fp.StudyTime))) # type is str
#print('type(rp_fp.RTPlanTime) = {}'.format(type(rp_fp.RTPlanTime))) # type is str
rp_fp.RTPlanTime= str(float(rs_fp.StudyTime) + 0.001)
rp_fp.InstanceCreationTime = str(float(rs_fp.InstanceCreationTime) + 0.001)


# rp_fp.RTPlanLabel = 'TEST'




# tandem_rp_line is made by information in the log of 
# run_and_make_rp_v02(folder='RP_Template_TestData', out_rp_filepath=r'out.brachy.rp.withpoints.v04.dcm')
# So it is tandem lin that for folder = RP_Template_TestData
# Case 5mm RP_Template_TestData
#tandem_rp_line = [[16.98905170843703, -184.51905170843702, 83.90804591543933], [16.12, -182.59439432365733, 79.43018494834773], [16.12, -180.22625544178416, 75.11276643836932], [15.61, -178.10368144202292, 70.66408125886848], [15.61, -175.89066242062734, 66.1888722377543], [15.1, -174.2040753078662, 61.57480512888706], [14.729683314929039, -172.1736095792839, 57.075705057419384], [14.58, -171.30564601520265, 52.373815716083655], [14.380884282801011, -169.63088428280102, 47.71915405020004], [14.07, -168.40476965158345, 42.91086137875857], [13.499484973334287, -167.15778337751826, 38.26268616993838], [13.56, -166.55221861146777, 33.68517102536384], [13.56, -165.3167540752058, 28.840212059630595], [13.05, -165.23, 23.915099979362182], [13.05, -164.02365616369758, 19.14787604601474], [12.54, -163.69, 14.292746333692945], [12.54, -163.18, 9.356747302684962], [12.03, -162.66623164553909, 4.4852221393690135], [12.03, -161.92495292378192, -0.4217532400709383], [12.03, -161.1751043098456, -5.326521885209219], [11.51, -160.62, -10.192257082467922], [11.51, -159.70766946316976, -15.077766811099007], [11.386390741412567, -159.09, -19.984742190538952], [11.0, -158.58, -24.872252237507816], [10.9244494605092, -158.06, -29.79627662545412], [10.49, -157.25898492424722, -34.64123559118737], [10.477706202490035, -156.53, -39.548210970627316], [9.98, -156.02, -44.421751807061376], [9.98, -155.51, -49.35775083806935], [9.98, -154.99, "-53.5"]]

# Case 4mm RP_Template_TestData
# tandem_rp_line = [[17.058034408590366, -184.83553835127827, 84.82143369180248], [16.12, -183.04872040645336, 80.32102040481051], [16.12, -180.4733483490366, 76.08175823151598], [15.61, -178.55800752481895, 71.55491671533126], [15.61, -176.28084178170548, 67.10357627515624], [15.110012332139032, -174.46002466427805, 62.53926404760405], [14.955327096954274, -172.61621853633338, 57.94356575751644], [14.789698894470218, -171.5756662234227, 53.30653420950084], [14.58, -169.87276039354526, 48.664463052097126], [14.07, -168.65186255883583, 43.87985317190524], [13.720887458553541, -167.6017749171071, 39.130931210013884], [13.521836263063076, -166.72183626306307, 34.64966171347814], [13.56, -165.56384698245822, 29.809203852777266], [13.05, -165.23, 24.915099979362207], [13.05, -164.48150629881926, 20.036905434600477], [12.735881996305874, -163.69, 15.26816469139559], [12.54, -163.18, 10.356747302684996], [12.273324552791502, -162.67, 5.454213932515696], [12.03, -162.16, 0.5487502096610601], [12.03, -161.42673816161056, -4.3586993784209165], [11.585182773532438, -160.6937369509645, -9.2108354864137], [11.51, -159.95476237042217, -14.108775017952333], [11.51, -159.21348364866498, -19.015750397392278], [11.0, -158.58, -23.87225223750779], [11.0, -158.06, -28.80575762101347], [10.49, -157.50607783149962, -33.6722437980407], [10.49, -156.76479910974243, -38.579219177480645], [9.98, -156.03933457348046, -43.424178143213894], [9.98, -155.51, -48.35775083806933], [9.98, -155.04252700099707, -53.29797307308827], [9.98, -154.99, "-53.5"]]

# Case 4mm in RALmilo
tandem_rp_line = [[-17.364760943476806, -137.50531764716507, 34.96922066277659], [-14.507295389260364, -138.96776102631532, 31.484506572268735], [-12.121197670078443, -141.8932334511995, 28.21556896195182], [-10.352067011648948, -143.68568033036905, 24.037129266612148], [-7.331626111466163, -145.72837388853384, 20.853412458988373], [-6.192040032243996, -146.9090550806545, 16.24906129029387], [-3.8038198803606917, -149.00737558337042, 12.431487658976454], [-2.747132679885409, -150.3128673201146, 7.870099029955784], [-1.0192642215390362, -151.67989966408996, 3.4381270994378488], [0.49, -153.63443097414736, -0.5117028736203919], [0.9360385068944075, -154.3260385068944, -5.351631839073963], [1.5000764344945428, -155.5407038230725, -10.125477715590891], [2.2818169860284883, -156.6618169860285, -14.920102945627201], [2.759779615212831, -157.45977961521285, -19.816846152805034], [2.9117725339140765, -158.2717725339141, -24.73801535705501], [2.79, -158.8, -29.651967982556688], [2.79, -159.55733472608358, -34.589907430809525], [2.79, -160.04573107506317, -39.549885303413205], [2.7110532270910443, -160.84894677290896, -44.478465290357306], [2.46, -161.33403799047878, -49.41841206350765], [2.5801821905215543, -161.76, -54.36418845612592], [3.12, -161.76, "-56"]]

import pydicom
import copy

# = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')


#print(rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].NumberOfControlPoints)
rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].NumberOfControlPoints = len(tandem_rp_line)
BCPItemTemplate = copy.deepcopy(rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence[0])
print('BCPItemTemplate = {}'.format(BCPItemTemplate))

rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence.clear()

# Clean Dose Reference
rp_fp. DoseReferenceSequence.clear()

# Start to make data style like these
"""
   (300a, 0112) Control Point Index                 IS: "0"
   (300a, 02d2) Control Point Relative Position     DS: "3.5"
   (300a, 02d4) Control Point 3D Position           DS: ['12.778672628366', '-14.486222523329', '1.80953906063796']
   (300a, 02d6) Cumulative Time Weight              DS: "0"
   ---------
   (300a, 0112) Control Point Index                 IS: "1"
   (300a, 02d2) Control Point Relative Position     DS: "3.5"
   (300a, 02d4) Control Point 3D Position           DS: ['12.778672628366', '-14.486222523329', '1.80953906063796']
   (300a, 02d6) Cumulative Time Weight              DS: "0"
   ---------
   (300a, 0112) Control Point Index                 IS: "2"
   (300a, 02d2) Control Point Relative Position     DS: "8.5"
   (300a, 02d4) Control Point 3D Position           DS: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
   (300a, 02d6) Cumulative Time Weight              DS: "0"
   ---------
   (300a, 0112) Control Point Index                 IS: "3"
   (300a, 02d2) Control Point Relative Position     DS: "8.5"
   (300a, 02d4) Control Point 3D Position           DS: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
   (300a, 02d6) Cumulative Time Weight              DS: "0"
   ---------
   (300a, 0112) Control Point Index                 IS: "4"
   (300a, 02d2) Control Point Relative Position     DS: "13.5"
   (300a, 02d4) Control Point 3D Position           DS: ['12.4938084644467', '-12.702280882149', '-8.025927609987']
   (300a, 02d6) Cumulative Time Weight              DS: "0"
   ---------
   (300a, 0112) Control Point Index                 IS: "5"
   (300a, 02d2) Control Point Relative Position     DS: "13.5"
   (300a, 02d4) Control Point 3D Position           DS: ['12.4938084644467', '-12.702280882149', '-8.025927609987']
   (300a, 02d6) Cumulative Time Weight              DS: "0"
"""
for idx, pt in enumerate(tandem_rp_line):
    BCPPt = copy.deepcopy(BCPItemTemplate)
    BCPPt.ControlPointRelativePosition = 3.5 + idx*5
    BCPPt.ControlPoint3DPosition[0] = pt[0]
    BCPPt.ControlPoint3DPosition[1] = pt[1]
    BCPPt.ControlPoint3DPosition[2] = pt[2]
    BCPStartPt = copy.deepcopy(BCPPt)
    BCPEndPt = copy.deepcopy(BCPPt)
    BCPStartPt.ControlPointIndex = 2*idx
    BCPEndPt.ControlPointIndex = 2*idx + 1 
    rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence.append(BCPStartPt)
    rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence.append(BCPEndPt)
    
    
    
    
# Change ROINumber of RP_Template_TestData RS into output RP output file 
#rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].ReferencedROINumber = 15
#rp_fp.ApplicationSetupSequence[0].ChannelSequence[1].ReferencedROINumber = 16
#rp_fp.ApplicationSetupSequence[0].ChannelSequence[2].ReferencedROINumber = 17


rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].ReferencedROINumber = 21
rp_fp.ApplicationSetupSequence[0].ChannelSequence[1].ReferencedROINumber = 22
rp_fp.ApplicationSetupSequence[0].ChannelSequence[2].ReferencedROINumber = 23

#print(BrachyControlPointSequenceItemTemplate)

#
# (ControlPointIndex, ControlPointRelativePosition, ControlPoint3DPosition, CumulativeTimeWeight)
# (0, 3.5, [,,], 0 )
# (0, )

#for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):






pydicom.write_file(out_rp_filepath, rp_fp)



attr=PhysiciansOfRecord, 
 In RS->Chen^Shang-Wen 
 In RP->Chen^Shang-Wen
after update, RP->Chen^Shang-Wen

attr=PatientName, 
 In RS->SUNARTI 
 In RP->李吳雲霞
after update, RP->SUNARTI

attr=PatientID, 
 In RS->34982640 
 In RP->21328949
after update, RP->34982640

attr=PatientBirthDate, 
 In RS->19761011 
 In RP->19620209
after update, RP->19761011

attr=PatientBirthTime, 
 In RS->000000 
 In RP->000000
after update, RP->000000

attr=PatientSex, 
 In RS->F 
 In RP->F
after update, RP->F

attr=DeviceSerialNumber, 
 In RS->417454940236 
 In RP->417454940236
after update, RP->417454940236

attr=SoftwareVersions, 
 In RS->13.6.32 
 In RP->13.6.32
after update, RP->13.6.32

attr=StudyID, 
 In RS->3 
 In RP->3
after update, RP->3

attr=StudyDate, 
 In RS->20190423 
 In RP->20191015
after update, RP->20190423

attr=StudyTime, 
 In RS->095710.840000 
 In RP->103307.760000
after update, RP->095710.840000

attr=StudyInstanceUID, 
 In RS->1.3.12.2.1107.5.1.4.95999.30000019042300043162800000016 
 In

# Check the data in output RP filepath 

In [54]:
rp_fp = pydicom.read_file(out_rp_filepath)
print(rp_fp)

#print(rp_fp.ApplicationSetupSequence[0].ChannelSequence[0])


(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20190423'
(0008, 0013) Instance Creation Time              TM: '181625.41999999998'
(0008, 0016) SOP Class UID                       UI: RT Plan Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.5.417454940236.2063186.20190423333506
(0008, 0020) Study Date                          DA: '20190423'
(0008, 0030) Study Time                          TM: '095710.840000'
(0008, 0050) Accession Number                    SH: '2019-10-15-10:36'
(0008, 0060) Modality                            CS: 'RTPLAN'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Plans'
(0008,

In [55]:
print(rs_fp)

(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20191017'
(0008, 0013) Instance Creation Time              TM: '181625.419000'
(0008, 0016) SOP Class UID                       UI: RT Structure Set Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.4.417454940236.268413.20191017181440
(0008, 0020) Study Date                          DA: '20190423'
(0008, 0030) Study Time                          TM: '095710.840000'
(0008, 0050) Accession Number                    SH: ''
(0008, 0060) Modality                            CS: 'RTSTRUCT'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Structure Sets'
(0008, 1

In [8]:




def get_new_uid(old_uid='1.2.246.352.71.5.417454940236.2063186.20191015164204', study_date='20190923'):
    def gen_6_random_digits():
        import random
        ret_str = ""
        for i in range(6):
            ch = chr(random.randrange(ord('0'), ord('9') + 1))
            ret_str += ch
        return ret_str

    theStudyDate = study_date
    uid_last_part = uid.split('.')[-1] #20191015164204  first 8 digit is match to StudyDate
    uid_list = uid.split('.')
    uid_list[-1] = theStudyDate + gen_6_random_digits()
    #print(uid_list)
    new_uid = '.'.join(uid_list)
    #print(uid)
    #print(new_uid)
    return new_uid


 




['1', '2', '246', '352', '71', '5', '417454940236', '2063186', '20190923456993']
1.2.246.352.71.5.417454940236.2063186.20191015164204
1.2.246.352.71.5.417454940236.2063186.20190923456993


In [60]:
rp_fp = pydicom.read_file('brachy.rp.dcm')
print(rp_fp)


(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20190423'
(0008, 0013) Instance Creation Time              TM: '181625.41999999998'
(0008, 0016) SOP Class UID                       UI: RT Plan Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.5.417454940236.2063186.20190423103304
(0008, 0020) Study Date                          DA: '20190423'
(0008, 0030) Study Time                          TM: '095710.840000'
(0008, 0050) Accession Number                    SH: '2019-10-15-10:36'
(0008, 0060) Modality                            CS: 'RTPLAN'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Plans'
(0008,

In [63]:
#rs_fp = pydicom.read_file(r'16568131/RS.1.2.246.352.71.4.417454940236.288142.20191018151404.dcm')
rs_fp = pydicom.read_file(r'24460566/RS.1.2.246.352.71.4.417454940236.288187.20191018155523.dcm')

print(rs_fp)

(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20191018'
(0008, 0013) Instance Creation Time              TM: '155540.373000'
(0008, 0016) SOP Class UID                       UI: RT Structure Set Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.4.417454940236.288187.20191018155523
(0008, 0020) Study Date                          DA: '20191008'
(0008, 0030) Study Time                          TM: '092849.803000'
(0008, 0050) Accession Number                    SH: '2019-10-08-09:42'
(0008, 0060) Modality                            CS: 'RTSTRUCT'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Structur